[Bugfix] Fix Kimi-K2 tool parser concatenated tool calls parsing

yiyeguhu opened the pull request:
## [Bugfix] Fix Kimi-K2 tool parser concatenated tool calls parsing

## Purpose
- Resolves GitHub issue #24478
- Fix regex pattern to handle concatenated tool calls without spacing
- Use negative lookahead to prevent over-matching across tool call boundaries
- Add comprehensive test cases for edge cases including angle brackets in JSON
- Consolidate test cases into parametrized test for better maintainability

## Test Plan
pytest tests/tool_use/test_kimi_k2_tool_parser.py

## Test Result
All 11 tests pass, including 4 new test cases that validate the regex fix:
- `concatenated_tool_calls_bug_fix`: Exact GitHub issue reproduction
- `three_concatenated_tool_calls`: Scalability test
- `mixed_spacing_tool_calls`: Robustness test  
- `angle_brackets_in_json`: Edge case with HTML in JSON

yiyeguhu committed:
[Bugfix] Fix Kimi-K2 tool parser concatenated tool calls parsing

- Fix regex pattern to handle concatenated tool calls without spacing
- Use negative lookahead to prevent over-matching across tool call boundaries
- Add comprehensive test cases for edge cases including angle brackets in JSON
- Consolidate test cases into parametrized test for better maintainability
- Resolves GitHub issue #24478

Signed-off-by: Thomas Mao <yiyeguhu@gmail.com>

chaunceyjiang reviewed (commented):
/cc @MoyanZitto PTAL.

yiyeguhu committed:
Add re.DOTALL flag to handle newlines in JSON arguments

- Add re.DOTALL flag to tool_call_regex to ensure . matches newlines
- Add test case for pretty-printed JSON with newlines
- Addresses reviewer feedback about newline handling in function arguments

Signed-off-by: Thomas Mao <yiyeguhu@gmail.com>

yiyeguhu commented on tests/tool_use/test_kimi_k2_tool_parser.py:
```diff
@@ -96,6 +100,78 @@ def test_extract_tool_calls_no_tools(kimi_k2_tool_parser):
+                         type='function'),
+                ToolCall(id='functions.send_email:2',
+                         function=FunctionCall(
+                             name="send_email",
+                             arguments='{"to": "user@example.com", "subject": "Daily Update"}',
+                         ),
+                         type='function')
+            ],
+            "I'll help you with multiple tasks. ",
+        ),
+        (
+            """Mixed spacing test. <|tool_calls_section_begin|> <|tool_call_begin|> functions.test:0 <|tool_call_argument_begin|> {} <|tool_call_end|><|tool_call_begin|>functions.test2:1<|tool_call_argument_begin|>{}<|tool_call_end|> <|tool_calls_section_end|>""",
+            [
+                ToolCall(id='functions.test:0',
+                         function=FunctionCall(
+                             name="test",
+                             arguments="{}",
+                         ),
+                         type='function'),
+                ToolCall(id='functions.test2:1',
+                         function=FunctionCall(
+                             name="test2",
+                             arguments="{}",
+                         ),
+                         type='function')
+            ],
+            "Mixed spacing test. ",
+        ),
+        (
+            """I need to process HTML content. <|tool_calls_section_begin|><|tool_call_begin|>functions.process_html:0<|tool_call_argument_begin|>{"html": "<div>content</div>", "text": "normal text"}<|tool_call_end|><|tool_calls_section_end|>""",
+            [
+                ToolCall(id='functions.process_html:0',
+                         function=FunctionCall(
+                             name="process_html",
+                             arguments='{"html": "<div>content</div>", "text": "normal text"}',
+                         ),
+                         type='function')
+            ],
+            "I need to process HTML content. ",
+        ),
```
OK. I found json.dump examples in test file.

yiyeguhu committed:
Replace hardcoded JSON strings with json.dumps in test cases

- Replace hardcoded JSON strings with json.dumps() calls for better maintainability
- Use default json.dumps format to match actual parser output (with spaces)

Signed-off-by: Thomas Mao <yiyeguhu@gmail.com>

yiyeguhu committed:
Merge branch 'main' into fix-kimi-k2-multiple-tool-calls-parsing-issue

yiyeguhu commented on vllm/entrypoints/openai/tool_parsers/kimi_k2_tool_parser.py:
```diff
@@ -38,7 +38,7 @@ def __init__(self, tokenizer: AnyTokenizer):
         self.tool_call_regex = re.compile(
-            r"<\|tool_call_begin\|>\s*(?P<tool_call_id>.+:\d+)\s*<\|tool_call_argument_begin\|>\s*(?P<function_arguments>.*?)\s*<\|tool_call_end\|>"
+            r"<\|tool_call_begin\|>\s*(?P<tool_call_id>[^<]+:\d+)\s*<\|tool_call_argument_begin\|>\s*(?P<function_arguments>(?:(?!<\|tool_call_begin\|>).)*?)\s*<\|tool_call_end\|>"
         )
```
Done

yiyeguhu committed:
Apply code formatting from pre-commit hooks

- Improve code formatting consistency
- Address pre-commit formatting requirements

Signed-off-by: Thomas Mao <yiyeguhu@gmail.com>

yiyeguhu committed:
Merge remote changes with local formatting updates

Signed-off-by: Thomas Mao <yiyeguhu@gmail.com>

chaunceyjiang committed:
Merge branch 'main' into fix-kimi-k2-multiple-tool-calls-parsing-issue

mehmetbaykar commented:
any update on this?

chaunceyjiang commented:
@bbartels Could you help submit a new PR based on this one?
